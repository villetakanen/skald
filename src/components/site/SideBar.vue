<template>
  <v-card class="sidebar">
    <v-card-text>
      <div v-html="content"></div>
    </v-card-text>
  </v-card>
</template>

<script lang="ts">
import { defineComponent, computed, ref, onMounted, watch } from '@vue/composition-api'
import Loading from '@/components/Loading.vue'
import { useSite } from '@/lib/useSite'
import Skaldmd from '@/lib/skaldmd'
import { Page } from '@/plugins/skaldfire'
import firebase from 'firebase/app'
import 'firebase/firestore'
import { update } from 'lodash'

export default defineComponent({
  components: {
    Loading
  },
  setup () {
    const { site } = useSite()
    const pagePrototype:Page = {
      pageid: '',
      siteid: '',
      name: '',
      content: '',
      htmlContent: '',
      htmlContentDraft: ''
    }
    const page = ref(pagePrototype)
    const loading = ref(false)
    let unsubscribe = () => {}
    let activeSiteid = ''
    const content = computed(() => {
      if (!page) return ''
      if (page.value.htmlContentDraft) return page.value.htmlContentDraft
      else if (page.value.htmlContent) return page.value.htmlContent
      const skaldmd = new Skaldmd(site.value.siteid)
      return skaldmd.toHtml(page.value.content)
    })
    function updateSidebar () {
      if (site.value.siteid === activeSiteid) return
      activeSiteid = site.value.siteid

      // Site changed, lets update sidebar
      const db = firebase.firestore()
      const pageRef = db.collection('sites').doc(activeSiteid).collection('pages').doc('sidebar')
      unsubscribe = pageRef.onSnapshot((pageSnapShot) => {
        page.value = {
          pageid: '',
          siteid: '',
          name: '',
          content: '',
          htmlContent: '',
          htmlContentDraft: ''
        }
        page.value.siteid = activeSiteid
        page.value.pageid = 'sidebar'
        page.value.name = pageSnapShot.data()?.name
        page.value.content = pageSnapShot.data()?.content
        page.value.htmlContent = pageSnapShot.data()?.htmlContent
        page.value.htmlContentDraft = pageSnapShot.data()?.htmlContentDraft
      })
    }
    onMounted(() => { updateSidebar() })
    watch(site, updateSidebar)
    return { loading, page, site, content }
  }
})
</script>
