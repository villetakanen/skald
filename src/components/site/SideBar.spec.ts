import { shallowMount } from '@vue/test-utils'
import SideBar from './SideBar.vue'

describe('Mounted App', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(SideBar, {
      mocks: {
        $t: () => {}
      },
      propsData: {
        siteid: 'mekanismi',
        pageid: 'ohjeita'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
