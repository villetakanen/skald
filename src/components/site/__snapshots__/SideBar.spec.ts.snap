// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounted App renders correctly 1`] = `
<v-card-stub
  class="sidebar"
  loaderheight="4"
  pageid="ohjeita"
  siteid="mekanismi"
  tag="div"
>
  <v-card-title-stub>
    
  </v-card-title-stub>
   
  <v-card-text-stub>
    <div />
  </v-card-text-stub>
   
  <!---->
</v-card-stub>
`;
